internal/provider/resource_password.go

```diff
@@ -1,22 +1,93 @@
 package provider
 
 import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"golang.org/x/crypto/bcrypt"
 )
 
 func resourcePassword() *schema.Resource {
+	passwordSchema := stringSchemaV1(true)
+	passwordSchema["bcrypt_hash"] = &schema.Schema{
+		Description: "A bcrypt hash of the generated random string.",
+		Type:        schema.TypeString,
+		Computed:    true,
+		Sensitive:   true,
+	}
+
+	create := func(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+		diags := createStringFunc(true)(ctx, d, meta)
+		if diags.HasError() {
+			return diags
+		}
+
+		if d.Get("bcrypt_hash") == "" {
+			hash, err := generateHash(d.Get("result").(string))
+			if err != nil {
+				diags = append(diags, diag.Errorf("err: %s", err)...)
+				return diags
+			}
+
+			if err := d.Set("bcrypt_hash", hash); err != nil {
+				diags = append(diags, diag.Errorf("err: %s", err)...)
+				return diags
+			}
+		}
+
+		return nil
+	}
+
 	return &schema.Resource{
 		Description: "Identical to [random_string](string.html) with the exception that the result is " +
 			"treated as sensitive and, thus, _not_ displayed in console output. Read more about sensitive " +
 			"data handling in the [Terraform documentation](https://www.terraform.io/docs/language/state/sensitive-data.html).\n" +
 			"\n" +
 			"This resource *does* use a cryptographic random number generator.",
-		CreateContext: createStringFunc(true),
+		CreateContext: create,
 		ReadContext:   readNil,
 		DeleteContext: RemoveResourceFromState,
-		Schema:        stringSchemaV1(true),
+		Schema:        passwordSchema,
 		Importer: &schema.ResourceImporter{
 			StateContext: importStringFunc(true),
 		},
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Version: 0,
+				Type:    resourcePasswordV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: resourcePasswordStateUpgradeV0,
+			},
+		},
 	}
 }
+
+func resourcePasswordV0() *schema.Resource {
+	return &schema.Resource{
+		Schema: stringSchemaV1(true),
+	}
+}
+
+func resourcePasswordStateUpgradeV0(ctx context.Context, rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	result, ok := rawState["result"].(string)
+	if !ok {
+		return nil, fmt.Errorf("resource password state upgrade failed, result could not be asserted as string: %T", rawState["result"])
+	}
+
+	hash, err := generateHash(result)
+	if err != nil {
+		return nil, fmt.Errorf("resource password state upgrade failed, generate hash error: %v", err)
```

Comment:
Conventionally after Go 1.13, it is recommended to use the `%w` (wrapped error) formatter for errors.

This does leave an interesting question though, what errors can be generated and are they recoverable? If they are not recoverable (e.g. cannot be fixed by retrying), practitioners will be forever stuck with an unusable resource unless they pin to an older provider version or the state upgrade code is "fixed" or updated to skip on the generation error.

Reply:
Have wrapped the error.

The errors that can be generated relate to:
- [checkCost](https://cs.opensource.google/go/x/crypto/+/46612604:bcrypt/bcrypt.go;l=138) - this should never return an error as we're using _DefaultCost_.
- [io.ReadFull](https://cs.opensource.google/go/x/crypto/+/46612604:bcrypt/bcrypt.go;l=145) - this will return an error if _maxSaltSize_ bytes cannot be read from _rand.Reader_.
- [bcrypt](https://cs.opensource.google/go/x/crypto/+/46612604:bcrypt/bcrypt.go;l=151) - can return error if underlying cipher funcs return error.